### PR TITLE
Add Callbacks interface

### PR DESCRIPTION
**Pull request by darthtrevino**
This PR adds callbacks to the ExecutionGraph for when nodes start/finish. This will allow us to centralize node execution and perf logging in client apps.

### CODE REVIEW

File: python/reactivedataflow/reactivedataflow/build_execution_graph.py [hunk 1]
```diff
@@ -23,7 +23,13 @@
 )
 from .execution_graph import ExecutionGraph
 from .model import Graph, Output, ValRef
-from .nodes import ExecutionNode, InputNode, Node
+from .nodes import (
```
Comment (AlonsoGuevara) [inline python/reactivedataflow/reactivedataflow/build_execution_graph.py:26]: For callbacks, wouldn't it be better to have them imported from nodes.callbacks?

File: python/reactivedataflow/reactivedataflow/definitions.py [hunk 2]
```diff
@@ -3,7 +3,14 @@
 
 from .config_provider import ConfigProvider
 from .decorators import AnyFn, Decorator
-from .nodes import EmitCondition, FireCondition, VerbFunction
+from .nodes import (
```
Comment (AlonsoGuevara) [inline python/reactivedataflow/reactivedataflow/definitions.py:6]: having everything at nodes level makes it a it confusing, may be my ignorance on reactivedataflow, but seems a bit odd

File: python/reactivedataflow/reactivedataflow/execution_graph.py [hunk 3]
```diff
@@ -16,20 +16,35 @@ class ExecutionGraph:
...
+        on_node_start: OnNodeStartCallback | None = None,
+        on_node_finish: OnNodeFinishCallback | None = None,
     ):
         """Initialize the execution graph.
 
         Args:
             nodes: The nodes in the graph.
             outputs: The outputs of the graph.
             order: The topological order of the nodes, starting with input nodes.
+            on_node_start: The callback for when a node starts.
```
Comment (AlonsoGuevara) [inline python/reactivedataflow/reactivedataflow/execution_graph.py:36]: One thing that doesn't clicks much with me is how much code is required to modify callbacks, or to add a new one if needed. Will sound cliche, but, have you tried with Protocols?

File: python/reactivedataflow/reactivedataflow/build_execution_graph.py (hunk 1 above)
Comment (darthtrevino) [reply python/reactivedataflow/reactivedataflow/build_execution_graph.py]: There's a fair amount of re-exporting happening in this library. `reactivedataflow.nodes` contains common node types and utility definitions. I've updated the PR to use more specific imports at these lines though.

File: python/reactivedataflow/reactivedataflow/definitions.py (hunk 2 above)
Comment (darthtrevino) [reply python/reactivedataflow/reactivedataflow/definitions.py]: As above, this has to due with the re-export structure here.

File: python/reactivedataflow/reactivedataflow/execution_graph.py (hunk 3 above)
Comment (darthtrevino) [reply python/reactivedataflow/reactivedataflow/execution_graph.py]: Yeah I think you have a point here. There are a few layers that these have to be wired through. I replaced individual callbacks with a top-level Callbacks protocol.

File: python/reactivedataflow/reactivedataflow/build_execution_graph.py [hunk 4]
```diff
@@ -176,7 +182,7 @@ def bind_inputs(
             node = nodes[nid]
             if isinstance(node, InputNode):
```
Comment (andresmor-ms) [inline python/reactivedataflow/reactivedataflow/build_execution_graph.py:183]: you could replace the ifs with structural pattern matching, you can do:

```py
match(node):
    case InputNode():
        node.attach(inputs[nid])
    case VerbNode():
        ...
    case _:
        pass
```

File: python/reactivedataflow/reactivedataflow/nodes/verb_node.py [hunk 5]
```diff
@@ -183,9 +195,29 @@ def _schedule_recompute(self, cause: str | None) -> None:
...
         result = await self._fn(inputs)
+        end = asyncio.get_event_loop().time()
+        self._fire_finish(end - start)
+
+        # Update the outputs
         if not result.no_output:
             for name, value in result.outputs.items():
                 self._output(name).on_next(value)
+                self._output(name).on_next(value)
+                self._output(name).on_next(value)
```
Comment (AlonsoGuevara) [inline python/reactivedataflow/reactivedataflow/nodes/verb_node.py:223]: Is this intended?

File: python/reactivedataflow/tests/unit/nodes/test_verb_node.py [hunk 6]
```diff
@@ -355,3 +344,4 @@ def on_output(o: str):
     node.attach(array_inputs=[rx.just(2), rx.just(4), rx.of(5, 6)])
     await node.drain()
     assert output == 12
+    assert output == 12
```
Comment (andresmor-ms) [inline python/reactivedataflow/tests/unit/nodes/test_verb_node.py:347]: I think this line is duplicated

File: python/reactivedataflow/reactivedataflow/nodes/verb_node.py (hunk 5 above)
Comment (darthtrevino) [reply python/reactivedataflow/reactivedataflow/nodes/verb_node.py]: Nope, that's weird, removed

File: python/reactivedataflow/tests/unit/nodes/test_verb_node.py (hunk 6 above)
Comment (darthtrevino) [reply python/reactivedataflow/tests/unit/nodes/test_verb_node.py]: Yep, weird - removed

File: python/reactivedataflow/reactivedataflow/build_execution_graph.py (hunk 4 above)
Comment (darthtrevino) [reply python/reactivedataflow/reactivedataflow/build_execution_graph.py]: done!

**Review by AlonsoGuevara — APPROVED** [review]

**Pull request merged**